guardian/editions
# Amazon deployment POC

## Why are you doing this?

Note: this is potentially not safe to merge, see "How to test"

This PR adds a new workflow to upload to the amazon store. It uses the community plugin [fastlane-plugin-amazon_app_submission](https://github.com/ugroupmedia/fastlane-plugin-amazon-app-store-submission), given that [there is no official one](https://github.com/fastlane/fastlane/issues/15148).

The plugin isn't perfect, in previous executions the amazon store API rejected the apk but the plugin did not surface that as an error and the build was green, but it seems that the functionality is working OK.

The authentication was setup [as described in the documentation](https://developer.amazon.com/docs/app-submission-api/auth.html) and is configured using three github secrets
 *  AMAZON_CLIENT_ID
 *  AMAZON_CLIENT_SECRET
 * AMAZON_APP_ID

It was unclear to me if the app id was an actual secret or just "private information", but I decided to define it as a github secret anyway.

## How to test

I was able to successfully upload build version 14001422 using this workflow ([see this run log](https://github.com/guardian/editions/actions/runs/6298241453)), but there's a detail that might make merging this workflow problematic.

![Screenshot 2023-09-25 at 12 05 28](https://github.com/guardian/editions/assets/1672034/a9a17c7e-73f7-4d88-9f88-3dec797e67d2)


The amazon upload functionality relies on importing a plugin, which had not been done before in this repo. Using plugins required setting up  `Pluginfile` and a `Gemfile`. This means that instead of relying on the fastlane executable provided by the GHA environment, this workflow relies on using the ruby cli `bundle` to install fastlane using `bundle install` and invoke fastlane using `bundle exec fastlane`. It's unclear to me if this breaks the existing workflows or not

So I recommend
- Merging this PR just before an android beta
- Ship an android beta
- If it fails, either fix forward or revert this branch

It would also be worthwhile testing if the APK built by this workflow is fully functioning before shipping it to users. The apk is available as a [build artifact here](https://github.com/guardian/editions/actions/runs/6298241453)
